planner/core/logical_plan_builder.go

```diff
@@ -4244,6 +4245,55 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	return result, nil
 }
 
+func (ds *DataSource) extractFD() *fd.FDSet {
+	// FD in datasource (leaf node) can be cached and reused.
+	if ds.fdSet == nil {
+		fds := &fd.FDSet{}
+		allCols := fd.NewFastIntSet()
+		// should use the column's unique ID avoiding fdSet conflict.
+		for _, col := range ds.TblCols {
+			// todo: change it to int64
+			allCols.Insert(int(col.UniqueID))
+		}
+		for _, idx := range ds.tableInfo.Indices {
+			keyCols := fd.NewFastIntSet()
+
+			// consider invisible index.
+			if !idx.Invisible {
+				continue
+			}
+			// consider prefix indexes.
+			allColIsNotNull := true
+			for _, idxCol := range idx.Columns {
+				if idxCol.Length != types.UnspecifiedLength {
```

Comment:
If the index with prefix-column has a unique constraint, we can also consider it actually?

Reply:
Ah, make sense. i'll remove the limitation here.
unique(char_column(10)), will also guarantee the prefix to be the unique which means the while column is unique too.